From d8a84dbac2db41e41a8c5601e66d8fe70031f2df Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Thu, 27 Jan 2022 11:06:57 +0200
Subject: [PATCH 63/64] tools: s32gen1image: Use different alignment for SD and
 QSPI boot

According to S32GEN1 Reference Manual the IVT should contain headers
aligned to 512 bytes when booting from SD and 8 bytes aligned when
booting from QSPI.

Issue: ALB-8432
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 tools/s32gen1image.c | 21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/tools/s32gen1image.c b/tools/s32gen1image.c
index a726729940..417b6606c2 100644
--- a/tools/s32gen1image.c
+++ b/tools/s32gen1image.c
@@ -34,6 +34,9 @@
 #define DCD_MAXIMUM_SIZE		(8192)
 #define DCD_HEADER_LENGTH_OFFSET	(1)
 
+#define BOOTROM_QSPI_ALIGNMENT		(0x8)
+#define BOOTROM_SD_ALIGNMENT		(0x200)
+
 #define DCD_COMMAND_HEADER(tag, len, params) ((tag) | \
 					      (cpu_to_be16((len)) << 8) | \
 					      (params) << 24)
@@ -148,17 +151,14 @@ static struct program_image image_layout = {
 	},
 	.dcd = {
 		.offset = S32_AUTO_OFFSET,
-		.alignment = 0x200U,
 		.size = DCD_MAXIMUM_SIZE,
 	},
 	.hse_reserved = {
 		.offset = S32_AUTO_OFFSET,
-		.alignment = 0x2000U,
 		.size = S32GEN1_SECBOOT_HSE_RES_SIZE,
 	},
 	.app_code = {
 		.offset = S32_AUTO_OFFSET,
-		.alignment = 0x200U,
 		.size = sizeof(struct application_boot_code),
 	},
 	.code = {
@@ -543,6 +543,20 @@ static void set_dcd_size(struct program_image *program_image)
 	program_image->dcd.size = iconfig.dcd.size;
 }
 
+static void set_headers_alignment(struct program_image *program_image)
+{
+	size_t alignment;
+
+	if (iconfig.flash_boot)
+		alignment = BOOTROM_QSPI_ALIGNMENT;
+	else
+		alignment = BOOTROM_SD_ALIGNMENT;
+
+	program_image->dcd.alignment = alignment;
+	program_image->hse_reserved.alignment = alignment;
+	program_image->app_code.alignment = alignment;
+}
+
 static int s32g2xx_build_layout(struct program_image *program_image,
 				size_t *header_size, void **image)
 {
@@ -557,6 +571,7 @@ static int s32g2xx_build_layout(struct program_image *program_image,
 	size_t last_comp = ARRAY_SIZE(parts) - 1;
 
 	set_dcd_size(program_image);
+	set_headers_alignment(program_image);
 
 	/* Compute auto-offsets */
 	s32_compute_dyn_offsets(parts, ARRAY_SIZE(parts));
-- 
2.17.1

